### PR TITLE
Fix the Fast Up-To-Date check for Roslyn.sln

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: build/Targets/Imports.targets [hunk 1]
```diff
@@ -2,7 +2,10 @@
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
 
-  
+  <PropertyGroup>
+    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+  </PropertyGroup>
```
Comment (jaredpar) [inline build/Targets/Imports.targets]: Why is this needed?

Comment (sharwell) [reply build/Targets/Imports.targets]: IIRC, changes to this file were not triggering an update to the build outputs. After a change was made to this file, its timestamp would end up newer than the output but CoreCompile wouldn't run so it would always be newer than the output. @rainersigwald should be able to provide more details about why this occurs.

Comment (rainersigwald) [reply build/Targets/Imports.targets]: I'm still not sure how it got into the FUTD inputs (@panopticoncentral?), but that's the reason things never got up to date--FUTD was considering it out of date because `Imports.targets` had been modified since the last time the compiler ran, so a build was run, but then CoreCompile skipped because it wasn't listed in its "try to account for property changes by rebuilding when build logic changes" input:

https://github.com/dotnet/roslyn/blob/469fc641a6bafe92a81e1efae05ff8153906ee99/src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets#L16-L17

**Comment by Pilchie** [comment]
Tagging @panopticoncentral too

File: build/Targets/Imports.targets (hunk 1 above)
Comment (panopticoncentral) [reply build/Targets/Imports.targets]: The current design of the up to date check is that we check to see if any of the imported target files have changed, assuming that if an imported target file has changed, the output might not be valid. This assumes that when we say its not up to date that MSBuild will agree and rebuild the project, so the outputs will be updated. If MSBuild isn't checking imported target files for up to date (and now that I'm writing this I'm kind-of remember having this discussion somewhere before), then I guess that logic is invalid and we should remove it since we'll get situations like this?

Comment (jaredpar) [reply build/Targets/Imports.targets]: What does this line do though?

Comment (sharwell) [reply build/Targets/Imports.targets]: @jaredpar it marks the file as an input for CoreCompile, forcing a compilation to complete if it changes.

Comment (jaredpar) [reply build/Targets/Imports.targets]: Is there any file for which this shouldn't be added? Trying to think of an example. Possible if a file was VS design time only I guess?

Comment (sharwell) [reply build/Targets/Imports.targets]: It's automatically added for **Directory.Build.props** and **Directory.Build.targets** for cases where the file is imported automatically by the project system. I'll add it for the others.

Comment (jaredpar) [reply build/Targets/Imports.targets]: Not quite what I'm asking. Still don't quite understand what this does. Trying to think of an example of a file where this wouldn't be needed. Will help me put this into context.

Comment (sharwell) [reply build/Targets/Imports.targets]: @jaredpar It's not needed for most files inside a NuGet package, because packages of a fixed version are immutable. If the contents of the package changed, it means the version of the package changed which results in a change to the generated NuGet .props/.targets in the Obj directory, which *do* have these lines.

Comment (rainersigwald) [reply build/Targets/Imports.targets]: @jaredpar It's needed in every imported MSBuild file and it should be done automatically and not manually. That is tracked as https://github.com/Microsoft/msbuild/issues/1299. For now, it should be done manually.

@panopticoncentral Yeah, I think we should consider backing out the import-tracking, or modifying it to be `$(MSBuildAllProjects)` based instead, to match the (bad) behavior of the build.

File: build/Targets/Roslyn.Toolsets.Xunit.targets [hunk 2]
```diff
@@ -2,6 +2,10 @@
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (jaredpar) [inline build/Targets/Roslyn.Toolsets.Xunit.targets]: In another file you removed the expanded `<Project>` syntax. This one remains.

File: build/Targets/Settings.props [hunk 3]
```diff
@@ -2,6 +2,10 @@
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (jaredpar) [inline build/Targets/Settings.props]: Same

File: build/Targets/Versions.props [hunk 4]
```diff
@@ -2,6 +2,10 @@
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (jaredpar) [inline build/Targets/Versions.props]: Same

File: build/Targets/Vsdconfig.targets [hunk 5]
```diff
@@ -1,6 +1,11 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (jaredpar) [inline build/Targets/Vsdconfig.targets]: Same

File: build/scripts/build.ps1 [hunk 6]
```diff
@@ -496,7 +496,7 @@ function Test-XUnit() {
 # Deploy our core VSIX libraries to Visual Studio via the Roslyn VSIX tool.  This is an alternative to 
 # deploying at build time.
 function Deploy-VsixViaTool() { 
-    $vsixDir = Get-PackageDir "roslyntools.microsoft.vsixexpinstaller"
+    $vsixDir = Get-PackageDir "RoslynTools.Microsoft.VSIXExpInstaller"
```
Comment (jaredpar) [inline build/scripts/build.ps1]: Why the casing change here? The lower case form is more apporpriate as it's how NuGet lays it out on disk.

Comment (sharwell) [reply build/scripts/build.ps1]: Get-VersionCore is now case sensitive against the package name as it exists in NuGet.

File: build/Targets/Roslyn.Toolsets.Xunit.targets (hunk 2 above)
Comment (sharwell) [reply build/Targets/Roslyn.Toolsets.Xunit.targets]: :arrow_right: The ones removed were required as part of the switch from PowerShell object syntax to `XmlNode.SelectSingleNode`. The alternative is managing a namespace table in the PowerShell code.

File: build/scripts/build.ps1 (hunk 6 above)
Comment (sharwell) [reply build/scripts/build.ps1]: If you want the names to match what appears on disk, we need to modify `Get-PackageDir` to correct the case in the invocation of `Join-Path`. Let me know if you want me to make this change.

Comment (jaredpar) [reply build/scripts/build.ps1]: Probaly should just change the implementation of Get-PackageDir to lower the case it's provided. That's the better fix and hides the implementation detail of NuGet.
